charms/worker/k8s/src/charm.py

```diff
@@ -1249,6 +1262,68 @@ def _on_refresh_certs_action(self, event: ops.ActionEvent):
             except (InvalidResponseError, K8sdConnectionError) as e:
                 event.fail(f"Failed to refresh certificates: {e}")
 
+    def _check_etcd_ready(self):
+        """Check if etcd is ready and update the status accordingly.
+
+        This method initializes the etcd instance and checks its readiness.
+        If etcd is not ready, it blocks the charm with an appropriate status.
+        """
+        if not BOOTSTRAP_DATASTORE.get(self) == DATASTORE_TYPE_EXTERNAL:
+            log.info("Not using external etcd, skipping external etcd readiness check")
+            return
+
+        legacy_etcd = self.model.get_relation(ETCD_RELATION)
+        charmed_etcd = self.model.get_relation(CHARMED_ETCD_RELATION)
+        etcd_certificate_relation = self.model.get_relation(ETCD_CERTIFICATES_RELATION)
+
+        if not legacy_etcd and not charmed_etcd:
+            msg = "Missing etcd relation"
+            log.error(msg)
+            status.add(ops.BlockedStatus(msg))
+            raise ReconcilerError(msg)
+
+        if legacy_etcd and charmed_etcd:
+            msg = "etcd and etcd-client relations are mutually exclusive"
+            log.error(msg)
+            status.add(ops.BlockedStatus(msg))
+            raise ReconcilerError(msg)
+
+        if charmed_etcd and not etcd_certificate_relation:
+            msg = "etcd-client relation requires etcd-certificates relation"
+            log.error(msg)
+            status.add(ops.BlockedStatus(msg))
+            raise ReconcilerError(msg)
+
+        if etcd_certificate_relation and legacy_etcd:
+            msg = "etcd-certificates relation is incompatible with etcd relation"
+            log.error(msg)
+            status.add(ops.BlockedStatus(msg))
+            raise ReconcilerError(msg)
+
+        if (
+            etcd_certificate_relation
+            and not self.etcd_certificates.certificates.get_assigned_certificates()[0]
+        ):
+            msg = "Waiting for the etcd client certificate"
+            log.error(msg)
+            status.add(ops.WaitingStatus(msg))
+            raise ReconcilerError(msg)
+
+    def _initialize_external_etcd(self) -> Union[EtcdReactiveRequires, CharmedEtcdRequires, None]:
```

Comment:
Then, we can type the return as the Protocol:
```suggestion
    def _initialize_external_etcd(self) -> Optional[EtcdRequiresProtocol]:
```